python/ray/train/callbacks/logging.py

```diff
@@ -1,112 +1,90 @@
+import json
+import logging
 import os
-from typing import Iterable, List, Optional, Dict, Set, Tuple, Union
-import abc
 import warnings
-import logging
-import numpy as np
-import json
 from pathlib import Path
+from typing import List, Optional, Dict, Set, Tuple, Union
+
+import numpy as np
 
-from ray.util.debug import log_once
-from ray.util.ml_utils.dict import flatten_dict
-from ray.util.ml_utils.json import SafeFallbackEncoder
 from ray.train.callbacks import TrainingCallback
+from ray.train.callbacks.results_prepocessors import IndexedResultsPreprocessor, \
+    ExcludedKeysResultsPreprocessor
+from ray.train.callbacks.results_prepocessors.preprocessor import \
+    SequentialResultsPreprocessor
 from ray.train.constants import (RESULT_FILE_JSON, TRAINING_ITERATION,
                                  TIME_TOTAL_S, TIMESTAMP, PID,
                                  TRAIN_CHECKPOINT_SUBDIR)
+from ray.util.debug import log_once
+from ray.util.ml_utils.dict import flatten_dict
+from ray.util.ml_utils.json import SafeFallbackEncoder
 from ray.util.ml_utils.mlflow import MLflowLoggerUtil
 
 logger = logging.getLogger(__name__)
 
 
-class TrainingLogdirCallback(TrainingCallback, abc.ABC):
-    """Abstract Train callback class with logging directory."""
-
-    def start_training(self, logdir: str, config: Dict, **info):
-        if self._logdir:
-            logdir_path = Path(self._logdir)
-        else:
-            logdir_path = Path(logdir)
+class TrainCallbackLogdirManager():
```

Comment:
Looking at this again, I think this manager class may be unnecessary. Can't this whole thing just be a utility function?

```python
def get_logdir_path(logdir: str, create_logdir_if_not_exists: bool) -> Path:
    ...
```

Then in `start_training` in the callbacks, you would just call this function `self._logdir = get_logdir_path(...)`, and make sure to pass in the correct logdir.

Reply:
I thought about this but this requires each caller to essentially define `logdir = logdir or default_logdir` themselves, in which case this method just contains logic for checking if the logdir exists.

 The class structure follows the actual call pattern more closely (`__init__` vs. `start_training`).

Reply:
Right, but that logic is simple enough that I think it's fine for the caller to do it themselves.

In my opinion, I think that the abstraction of passing in 2 logdirs to a manager is less intuitive than just figuring out the correct logdir in the callback itself, especially since it can literally just be 1 line.

Reply:
Yeah that's fair, but in that case is there any point in providing any utility method at all?

I will probably need to think about this some more. By changing it from a Mixin/TrainingCallback  to this new class, we did lose some of structure that glued it all together. Not so sure what the right abstraction is anymore...

Reply:
I think in the long run we may want to end up with something like what @Yard1 suggested:
> Do you think it would make sense to extend the concept of `ResultPreprocessor` to `start_training` as well? That would cover the logdir stuff. Not sure if the effort would be justifiable, though

While this could fit the pattern quite nicely, the desired interface would be a lot more clear when a second use-case comes around.